src/lib.rs

```diff
@@ -497,8 +501,8 @@ where
     DB: plumbing::GetQueryTable<Q>,
     Q: Query<DB> + 'me,
 {
-    db: &'me DB,
-    storage: &'me Q::Storage,
+    db: &'me mut DB,
+    storage: Arc<Q::Storage>,
```

Comment:
Hm, this encoding seems somewhat unoptimal now. Like we already have `&mut DB`, and that's strictly more than `Arc<Q::Storage>`. I wonder if we should change `InputQueryStorageOps::set` to not take `DB` as an argument. 

However, this is internal impl detail, so I am ok with landing this as is.

Reply:
Hmm, interesting thought. If we didn't have a `DB`, but only a runtime, it might work. However, it would prevent us from doing `salsa_event` calls -- at least from within the `InputStorage`. I'll take a look later. If we could do away with the `Arc` and `db` argument, it would mean that the storage could take `&mut self`, which could allow us to eliminate the mutex -- that seems kind of interesting.

Reply:
OK so it's going to be a bit tricky. The problem is this: *somebody* needs to be able to give us a "simultaneously live" `&mut Q::Storage` and `&mut SalsaRuntime`. Unfortunately, right now, those two bits of data are supplied by two distinct traits.

One thing that might help us here would be if we gave up the idea that the user defines the Salsa database type. Instead, salsa database would be a struct that **we** define -- but it has a "userdata" field that users can specify.

I've thought about this for other reasons. In particular, for the purposes of guaranteeing safety, it seems suboptimal that the DB -> runtime link is defined by end-users, who could do all manor of crazy things in there.

Reply:
(However that also seems out of scope for this PR)

Reply:
My inclination is to spin out my previous comment into an issue to be explored shortly, but land this PR in the meantime... agree, @matklad?

Reply:
:+1: